kedro/io/data_catalog.py

```diff
@@ -565,13 +616,33 @@ def list(self, regex_search: Optional[str] = None) -> List[str]:
             ) from exc
         return [dset_name for dset_name in self._data_sets if pattern.search(dset_name)]
 
+    def remove_pattern_matches(self, dataset_list: Set[str]):
+        """Helper method that checks which dataset names match a pattern in the catalog.
+        It returns a copy of the original list minus all those matched dataset names."""
+        if self.dataset_patterns:
+            dataset_list_minus_matched = []
+            for dataset in dataset_list:
+                # If dataset matches a pattern, remove it from the list.
+                for dataset_name in self.dataset_patterns.keys():
+                    result = parse(dataset_name, dataset)
+                    if result:
+                        break
+                    else:
+                        dataset_list_minus_matched.append(dataset)
+            return set(dataset_list_minus_matched)
+        return dataset_list
```

Comment:
This is much nicer:
```suggestion
        if not self.dataset_patterns:
                    return dataset_list

        dataset_list_minus_matched = dataset_list.copy()
        for dataset in dataset_list:
            matches = (parse(pattern, dataset) for pattern in self.dataset_patterns.keys())
            # If dataset matches any pattern, remove it from the list.
            if any(matches):
                dataset_list_minus_matched -= dataset
        return set(dataset_list_minus_matched)
```

As a general rule:

1. The shorter case in an `if` statement should **always** go first
2. If there's return, there's no need of `else` (like in the original code)
3. There's usually a nicer solution than using a `break` in a loop (not always though)